### PR TITLE
swap, min, and max

### PR DESCRIPTION
**Pull request by ibaned**

### CODE REVIEW

File: core/src/impl/Kokkos_Utilities.hpp [hunk 1]
```diff
@@ -411,6 +411,16 @@ struct inclusive_scan_integer_sequence
 
 }} // namespace Kokkos::Impl
 
+namespace Kokkos {
+
+template <class T>
+KOKKOS_INLINE_FUNCTION void swap(T& a, T& b) {
+  T c{std::move(a)};
+  a = std::move(b);
+  b = std::move(c);
```
Comment (dalg24) [inline core/src/impl/Kokkos_Utilities.hpp]: Shouldn't you be using `Kokkos::move()`?

Comment (ibaned) [reply core/src/impl/Kokkos_Utilities.hpp]: `std::move` actually works in device code with recent CUDA versions, not sure if we still support a CUDA version where it doesn't work. I can check on that.

Comment (mhoemmen) [reply core/src/impl/Kokkos_Utilities.hpp]: How recent is recent?

**Comment by crtrott** [comment]
Needs a unit test. Its fine to put that into the DefaultExecutionSpace tests.

File: core/src/impl/Kokkos_Utilities.hpp (hunk 1 above)
Comment (ibaned) [reply core/src/impl/Kokkos_Utilities.hpp]: I put `std::move` in `Kokkos::View` a while ago, so basically its required now. Our tests are still passing on CUDA 8 and 9 versions.

**Comment by ibaned** [comment]
Added `Kokkos::min` and `Kokkos::max`, also added a unit test for all three.
spot check:
```
Running on machine: sems
Repository Status:  653ef8334adcdce3672ce92db75c1761944a4fd5 Fix some build errors


Going to test compilers:  gcc/5.3.0 gcc/6.1.0 intel/17.0.1 clang/3.9.0 cuda/8.0.44
Testing compiler gcc/5.3.0
  Starting job gcc-5.3.0-OpenMP-release
  PASSED gcc-5.3.0-OpenMP-release
Testing compiler gcc/6.1.0
  Starting job gcc-5.3.0-OpenMP-hwloc-release
  PASSED gcc-5.3.0-OpenMP-hwloc-release
  Starting job gcc-6.1.0-Serial-release
  PASSED gcc-6.1.0-Serial-release
Testing compiler intel/17.0.1
  Starting job gcc-6.1.0-Serial-hwloc-release
  PASSED gcc-6.1.0-Serial-hwloc-release
  Starting job intel-17.0.1-OpenMP-release
  PASSED intel-17.0.1-OpenMP-release
Testing compiler clang/3.9.0
  Starting job intel-17.0.1-OpenMP-hwloc-release
  PASSED intel-17.0.1-OpenMP-hwloc-release
  Starting job clang-3.9.0-Pthread_Serial-release
  PASSED clang-3.9.0-Pthread_Serial-release
Testing compiler cuda/8.0.44
  Starting job clang-3.9.0-Pthread_Serial-hwloc-release
  PASSED clang-3.9.0-Pthread_Serial-hwloc-release
  Starting job cuda-8.0.44-Cuda_OpenMP-release
  PASSED cuda-8.0.44-Cuda_OpenMP-release
#######################################################
PASSED TESTS
#######################################################
clang-3.9.0-Pthread_Serial-hwloc-release build_time=200 run_time=181
clang-3.9.0-Pthread_Serial-release build_time=191 run_time=393
cuda-8.0.44-Cuda_OpenMP-release build_time=552 run_time=522
gcc-5.3.0-OpenMP-hwloc-release build_time=195 run_time=94
gcc-5.3.0-OpenMP-release build_time=190 run_time=166
gcc-6.1.0-Serial-hwloc-release build_time=223 run_time=236
gcc-6.1.0-Serial-release build_time=151 run_time=213
intel-17.0.1-OpenMP-hwloc-release build_time=500 run_time=156
intel-17.0.1-OpenMP-release build_time=434 run_time=166
```

**Comment by ibaned** [comment]
@crtrott can you take a look at this?

**Comment by ibaned** [comment]
addresses #1567

**Pull request merged**